### PR TITLE
[Flow] Move elementwise op fusion and bubble up expand shapes patterns into their own pass.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Currently these were run within the `FusionOfTensorOps` pass as a single fixed point. This makes it hard to debug what exactly is happening with fusion. Having them at separate pass boundaries allows for better debugability (especially on large models).

This pass should mostly be an NFC and is the first step towards breaking up `FusionOfTensors` pass.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/ElementwiseOpFusion.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace {
+
+class ElementwiseOpFusionPass
+    : public impl::ElementwiseOpFusionPassBase<ElementwiseOpFusionPass> {
+
+public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<affine::AffineDialect, linalg::LinalgDialect>();
+  }
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/ElementwiseOpFusion.cpp]: nit: Use the tablegen `dependentDialects` feature to register dialects.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp [hunk 2]
```diff
@@ -344,123 +344,6 @@ void FusionOfTensorOpsPass::runOnOperation() {
...
-          operands.insert(consumer->operand_begin(),
-                          std::next(consumer->operand_begin(),
-                                    fusedOperand->getOperandNumber()));
-          operands.insert(std::next(consumer->operand_begin(),
-                                    fusedOperand->getOperandNumber() + 1),
-                          consumer->operand_end());
-          if (operands.size() >= kIreeMaxOperandCount)
-            return false;
-
-          return areFusableOps(context, fusedOperand, fuseMultiReduction);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:376]: I do not see a use for this function anymore. It looks like it got moved and renamed to `FusionUtils`, but is still in this file?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td [hunk 3]
```diff
@@ -212,6 +220,19 @@ def DumpDispatchGraphPass : Pass<"iree-flow-dump-dispatch-graph-pass"> {
...
+  ];
+  let dependentDialects = [
+    "mlir::affine::AffineDialect",
+  ];
+}
+
+
 def ExportBenchmarkFuncsPass :
     Pass<"iree-flow-export-benchmark-funcs-pass", "mlir::ModuleOp"> {
   let summary = "Exports benchmark functions";
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td:238]: Can you add a description to `FusionOfTensorOpsPass`? After separating elementwise fusion and fusion by expansion that pass seems to mostly be focused on multi-use fusion and is no longer reflected by the name of the pass. Or change the name of the pass maybe. I noticed that a number of Flow passes could use better names (in particular the many dimension collapsing passes).

**Review by qedawkins — APPROVED** [review]
It looks mostly NFC so looks like a good cleanup to me. LGTM after addressing comments.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td (hunk 3 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td:238]: There are a few things to unwind with that pass still. I need to
1) Remove constant folding patterns there
2) Remove the collapse shape propagation patterns from there as well.

Once I get through 1 and 2, ill rename the `FusionOnTensors` pass.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/ElementwiseOpFusion.cpp (hunk 1 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/ElementwiseOpFusion.cpp]: ah, I actually did do that. Thanks!

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:376]: Ah, looks like I forgot to delete. Good catch!

**Pull request merged**